File: Algorithm_334_Increasing Triplet Subsequence.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    bool increasingTriplet(vector<int>& nums) {
        if (nums.size() < 3) return false;
        if (nums.size() == 3) return nums[0] < nums[1] && nums[1] < nums[2];
        vector<int> mins(nums.size());
        mins[0] = nums[0];
        for (int i = 1; i < nums.size() - 2; i++) mins[i] = min(mins[i - 1], nums[i]);
        for (int i = nums.size() - 2; i >= 1; i--)
        {
            if (mins[i - 1] < nums[i] && nums[i + 1] > nums[i]) return true;
            nums[i] = max(nums[i + 1], nums[i]);
        }
        return false;
    }
};

File: Algorithm_334_Increasing Triplet Subsequence.java
// 从左到右遍历数组nums，用small和big分别记录遇到的最小的值和次小的值
// 若small和big都已经被更新过，而此时又遇到一个比big更大的值
// 表明成功找到了所要求的subsequence，返回true
// 注意在返回true时，可能small的值已经发生了改变，例如nums = [1, 3, 0, 5]
// 在遇到5时，small = 0, big = 3，但是程序的结果不会受到影响
// 这是由于在更新big时，small的值为1，是符合要求的

class Solution {
    public boolean increasingTriplet(int[] nums) {
        int small = Integer.MAX_VALUE, big = Integer.MAX_VALUE;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] <= small) {
                small = nums[i];
            } else if (nums[i] <= big) { // nums[i] > small and <= big
                big = nums[i];
            } else { // successfully find a num that is >= small and big
                return true;
            }
        }
        return false;
    }
}